### PR TITLE
Provide Factory Methods for Power MemoryReference objects

### PR DESCRIPTION
**Pull request by midronij**
This PR does the following:

-Add static helper methods for each MemoryReference constructor
-Replace all direct calls to MemoryReference constructors with corresponding static helpers

### CODE REVIEW

**Review by aviansie-ben — COMMENTED** [review]
Just a quick comment before I start a full review: it looks like these changes are structured in such a way that they'll require a coordinated merge between OMR and OpenJ9. Perhaps this can be split into 2 PRs, with one to add the new static factory methods and switch OMR to use them and another to remove/privatize the old constructors and remove the int parameter hack? That should hopefully allow us to avoid needing a coordinated merge.

Barring that, somebody else will need to be the committer to merge this PR, as I'm not a committer at OpenJ9 and thus can't handle a coordinated merge between the two projects on my own.

**Comment by midronij** [comment]
@aviansie-ben I'll try splitting this PR in two to simplify the merging process

**Review by aviansie-ben — CHANGES_REQUESTED** [review]
Please rewrite the commit message subject line on the 2nd commit (e9d9f3f8fea8215c636a94d234f7266e1134d15b) to use the imperative mood (i.e. don't use past tense) per the [contribution guidelines](https://github.com/eclipse/omr/blob/master/CONTRIBUTING.md#commit-guidelines).

I see that a couple of changes in `OMRTreeEvaluator.cpp` seem to have escaped into #5537, which should be moved back into this PR. I can also still see some calls to the `TR::MemoryReference` constructors in the compiler unit tests that need to be updated.

File: compiler/p/codegen/PPCSystemLinkage.cpp [hunk 1]
```diff
@@ -714,11 +714,12 @@ TR::PPCSystemLinkage::createPrologue(
       if (size > (-LOWER_IMMED))
          {
          cursor = loadConstant(cg(), firstNode, -size, gr11, cursor);
-         cursor = generateMemSrc1Instruction(cg(),TR::InstOpCode::Op_stux, firstNode, new (trHeapMemory()) TR::MemoryReference(sp, gr11, TR::Compiler->om.sizeofReferenceAddress(), cg()), sp, cursor);
+         cursor = generateMemSrc1Instruction(cg(),TR::InstOpCode::Op_stux, firstNode, TR::MemoryReference::createWithIndexReg(cg(), sp, gr11, TR::Compiler->om.sizeofReferenceAddress()), sp, cursor);
          }
       else
          {
-         cursor = generateMemSrc1Instruction(cg(),TR::InstOpCode::Op_stu, firstNode, new (trHeapMemory()) TR::MemoryReference(sp, -size, TR::Compiler->om.sizeofReferenceAddress(), cg()), sp, cursor);
+         //NOTE (remove this comment later): the fact that the displacement here is negative might mess things up when displacement is changed from 32 bits to 64 bits
```
Comment (aviansie-ben) [inline compiler/p/codegen/PPCSystemLinkage.cpp:721]: This comment is identical to one I pointed out in my previous review and should also be removed.

**Review by aviansie-ben — CHANGES_REQUESTED** [review]

**Review by aviansie-ben — APPROVED** [review]

**Comment by aviansie-ben** [comment]
@genie-omr build plinux,aix

**Comment by aviansie-ben** [comment]
Looks like the compiler unit tests are failing to build with this change:

```
[2020-09-11T17:22:08.627Z] In file included from /home/jenkins/workspace/Build/fvtest/compilerunittest/p/Peephole.cpp:23:0:
[2020-09-11T17:22:08.628Z] /home/jenkins/workspace/Build/fvtest/compilerunittest/p/Util.hpp: In member function ‘TR::MemoryReference* MemoryReference::reify(TR::CodeGenerator*) const’:
[2020-09-11T17:22:08.628Z] /home/jenkins/workspace/Build/fvtest/compilerunittest/p/Util.hpp:94:13: error: expected primary-expression before ‘)’ token
[2020-09-11T17:22:08.628Z]              );
```

**Comment by aviansie-ben** [comment]
@genie-omr build plinux,aix

**Comment by aviansie-ben** [comment]
Not waiting for Travis since this is a Power-specific change

**Pull request merged**